writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
+			return build.ExtendErr("Failed to apply update", err)
+		}
+		// Remember new checksums to be able to remove unnecessary setup files
+		// later
+		checksums[hex.EncodeToString(su.ncs[:])] = struct{}{}
 	}
-	err = ca.applyUpdates(updates)
-	if err != nil {
-		return nil, err
+
+	// Sync the applied updates
+	if err := file.Sync(); err != nil {
+		return build.ExtendErr("Failed to sync database", err)
 	}
 
-	// Load the file into the array.
-	fileBytes, err := ioutil.ReadAll(file)
+	// Remove unnecessary setup files
+	files, err := ioutil.ReadDir(testdir)
 	if err != nil {
-		return nil, err
-	}
-	for i := 0; i+7 < len(fileBytes); i += 8 {
-		count := binary.LittleEndian.Uint64(fileBytes[i : i+8])
-		ca.countdown = append(ca.countdown, count)
+		return build.ExtendErr("Failed to get list of files in testdir", err)
 	}
-	if len(ca.countdown) == 0 {
-		// Count must at least start at zero.
-		err = ca.addCount()
-		if err != nil {
-			return nil, err
+	for _, f := range files {
+		_, exists := checksums[f.Name()]
+		if len(f.Name()) == 32 && !exists {
+			if err := deps.remove(filepath.Join(testdir, f.Name())); err != nil && !os.IsNotExist(err) {
+
+				return build.ExtendErr("Failed to remove setup file", err)
+			}
 		}
-		return ca, nil
 	}
-	start := ca.countdown[0]
-	if uint64(len(ca.countdown)) < start+1 {
-		return nil, errors.New("count is incorrect length")
-	}
-	for i, num := range ca.countdown {
-		if num == 0 {
-			ca.countdown = ca.countdown[:i+1]
-			break
+
+	// Check if the checksums match the data
+	numbers := make([]byte, numSilos*int64(numIncrease)*4)
+	var cs checksum
+	for _, silo := range silos {
+		// Adjust the size of numbers
+		numbers = numbers[:4*len(silo.numbers)]
+
+		// Read numbers and checksum
+		if _, err := silo.f.ReadAt(numbers, silo.offset); err != nil {
+			return build.ExtendErr("Failed to read numbers of silo", err)
 		}
-		if num != start-uint64(i) {
-			return nil, fmt.Errorf("count is incorrect representation %v != (%v - %v)", num, start, uint64(i))
+		if _, err := silo.f.ReadAt(cs[:], silo.offset+int64(4*len(silo.numbers))); err != nil {
+			return build.ExtendErr("Failed to read checksum of silo", err)
 		}
-	}
 
-	// Check that the splotch file follows the splotch rules.
-	splotchData, err := ioutil.ReadAll(ca.splotchFile)
-	if err != nil {
-		return nil, err
-	}
-	skip := 0
-	for i := 0; i < len(splotchData); i += skip {
-		skip += 80
-		current := binary.LittleEndian.Uint64(splotchData[i : i+8])
-		if current == 0 {
-			continue
-		}
-		for j := 8; j < skip; j += 8 {
-			next := binary.LittleEndian.Uint64(splotchData[j+i : j+i+8])
-			if next != current-1 {
-				return nil, fmt.Errorf("splotch does not count down correctly: %v != (%v - %v)", current, next, 1)
-			}
-			current = next
+		// The checksum should match
+		c := blake2b.Sum256(numbers)
+		if bytes.Compare(c[:checksumSize], cs[:]) != 0 {
+			return errors.New("Checksums don't match")
 		}
 	}
```

Comment:
I think you need an additional check here to make sure that the numbers are following the desired pattern - that they strictly increment except for one place where they decrement

Reply:
If they followed the pattern while the checksum was created they should also follow it when the checksum matches